lib/functions/installs.bash

```diff
@@ -117,30 +117,199 @@ install_local_tool_versions() {
+    display_debug "--------------------------------------------------------------------------------------------------------------"
+    tools_installed=$(install_directory_tools "$search_path" "$plugins_installed" "$tools_installed")
+    display_debug "install_directory_tools_recursive '$search_path': install_directory_tools returned tools_installed='$tools_installed'"
+
+    # terminate if all tools are installed
+    if [[ -n $(stringlist_a_subset_of_b "$plugins_installed" "$tools_installed") ]]; then
+      display_debug "install_directory_tools_recursive '$search_path': exiting because all known tools ($plugins_installed) are installed ($tools_installed)"
+      printf "%s\n" "$tools_installed"
+      return 0
+    fi
+
+    # go up a directory
+    search_path=$(dirname "$search_path")
+  done
+
+  # still got some tools to install...
+  # lets see if $ASDF_DEFAULT_TOOL_VERSIONS_FILENAME as actually
+  # an absolute path
+  if [ -f "$ASDF_DEFAULT_TOOL_VERSIONS_FILENAME" ]; then
+    display_debug "--------------------------------------------------------------------------------------------------------------"
+    display_debug "attempting to treat \$ASDF_DEFAULT_TOOL_VERSIONS_FILENAME as an absolute path: $ASDF_DEFAULT_TOOL_VERSIONS_FILENAME"
+    tools_installed=$(_install_directory_tools "" "$ASDF_DEFAULT_TOOL_VERSIONS_FILENAME" "$plugins_installed" "$tools_installed")
+    display_debug "install_directory_tools_recursive '$ASDF_DEFAULT_TOOL_VERSIONS_FILENAME': install_directory_tools returned tools_installed='$tools_installed'"
+  fi
+
+  printf "%s\n" "$tools_installed"
+}
+
+install_directory_tools() {
+  local search_path=$1
+  local plugins_installed=$2
+  local tools_installed=$3
+  display_debug "install_directory_tools '$search_path': starting install. tools_installed='$tools_installed'"
+
+  # install tools from .tool-versions
+  # install order is the order listed in .tool-versions
+  file_name=$(asdf_tool_versions_filename)
+  tools_installed=$(_install_directory_tools "$search_path" "$file_name" "$plugins_installed" "$tools_installed")
+
+  # install tools from legacy version files
+  # install order is plugin order which is alphabetical
+  local legacy_config
+  legacy_config=$(get_asdf_config_value "legacy_version_file")
+  if [ "$legacy_config" = "yes" ]; then
+    tools_installed=$(_install_directory_tools_legacy "$search_path" "$plugins_installed" "$tools_installed")
+  fi
+
+  printf "%s\n" "$tools_installed"
+}
+
+_install_directory_tools() {
+  local search_path=$1
+  local file_name=$2
+  local plugins_installed=$3
+  local tools_installed=$4
+
+  local tool_versions
+  if ! [[ -f "$search_path/$file_name" ]]; then
+    display_debug "_install_directory_tools '$search_path': exiting early... $file_name file not found"
+    printf "%s\n" "$tools_installed"
+    return 0
+  fi
+
+  tool_versions=$(strip_tool_version_comments "$search_path/$file_name" | awk '{$1=$1};1')
```

Comment:
Probably want this `$()` to be quoted right?

Reply:
I am no expert on bash and you could be right. I just found a lot of the time in installs.bash, there aren't many lines with `="$(...)"`. Should they be quoted ?

Reply:
Typically if the value produced by the expression contains spaces and you want it to stay as a string you should quote - `string_with_spaces="$(echo "string with spaces")"`. I think that is the case here, but I may be wrong. Generally, quoting is safer in Bash (unless you are dealing with an array, in which case it won't work).